### ISSUE TITLE
Huge memory leaks

### ISSUE THREAD

**Issue by vitakot** [Prototik/HoloEverywhere#432]
Hi,

I made a test with the MAT tool and found there are memory leaks. I have just created a trivial application using template in Eclipse, changed the base Activity to org.holoeverywhere.app.Activity and set:

android:theme="@style/Holo.Theme.NoActionBar"

in the manifest.

Then I started/finished this application 10 times. The result is that the heap grows by 175 kB, and the number of allocated objects increases by 3000 per each 10 repetitions.

I think it is a huge memory leak which can be cause of OOM exceptions, especially on low-end devices with old OS versions.

Device: Nexus 7, Android 4.2.2

**Comment by Prototik** [comment]
This object will be remove by garbage collector, no?
I'm also see 60k objects on startup, 70k after screen rotation, but after gc again 60k.

**Comment by vitakot** [comment]
Unfortunately not. I do not use screen rotation, just starting the app, pressing back and then call GC explicitly in the Eclipse DDMS view. After each such cycle memory allocation grows by several KB and never goes back, the exact amount depends on the actual device, screen size/density and the layout complexity.

I am testing it on Nexus 7 (Android 4.2.2, 1280 x 800 px). I have a simple login activity with three buttons on a black background, after each start/finish cycle the memory heap grows by 70 kB and the number of allocated object increases by approx. one thousand.

Just to note, previous version (from about two month ago) did not leak so much. Yesterday I updated to the latest revision and found out that the memory leakage has significantly increased.

**Comment by vitakot** [comment]
I have created a sample project using Google maps v2 API, it is in public SVN repository on Assembla:

https://www.assembla.com/code/gmaps_v2_mem_leak/subversion/nodes

There are a really huge memory leaks in conjunction with your library. After the first application start there is about 4 MB and 70 000 objects allocated in the VM heap, after five cycles of back/start it raises to 7 MB and 100 000 objects. Such a leak inevitably leads to OOM exception and app kill in a short time.

If I use native android.support.v4.app.FragmentActivity instead of your org.holoeverywhere.app.Activity, there are no memory leaks. More precisely, there are also leaks, but only these which introduces maps API as such – it is known and confirmed bug: http://code.google.com/p/gmaps-api-issues/issues/detail?id=4766

It looks I have no option than not to use holoeverywhere library and to create necessary styles on my own.

Testing device: LG-P970, Android 2.3.4

**Comment by httpdispatch** [comment]
I can confirm that leaks are present. Try HE demo and rotate screen few times. Number of objects and used memory keep growing even after gc

**Comment by httpdispatch** [comment]
Hope this will be useful.

I've tried abs-styled demo. Rotated screen few times. There are small leaks.
Then i've adjusted it to use Holo Activity and Fragment and application. And after that it become leaking 1000 times more. 2000-3000 objects + on each screen rotation

**Comment by httpdispatch** [comment]
Even simplier.

Create standard android project with one activity containing relativelayout and textview (default project), and adjust it to use HE. Memory leak is there

**Comment by httpdispatch** [comment]
Can it be problem in internal.WindowDecorView. How i see it has static members and number of views held by it grows on each screen rotation
First run
![image](https://f.cloud.github.com/assets/2315223/475854/f91bc198-b7a4-11e2-9284-1ef3fd7b3766.png)

Rotated screen
![image](https://f.cloud.github.com/assets/2315223/475852/dd35958a-b7a4-11e2-9adb-e2489ac03158.png)

**Comment by httpdispatch** [comment]
WindowDecorView: Path to GC root
![image](https://f.cloud.github.com/assets/2315223/475888/980d18fa-b7a6-11e2-9804-3a34e790bdbd.png)

**Comment by httpdispatch** [comment]
Ok, seems that cause is in mStatesMap of IAddon class. It is WeakHashMap. Keys are weakreferences but not values. To fix memory leak we can use such approach

``` java
@SuppressWarnings("unchecked")
    public <T, V extends IAddonBase<T>> V obtain(T object) {
        try {
            WeakReference<V> addonRef = (WeakReference<V>) mStatesMap
                    .get(object);
            V addon = addonRef == null ? null : addonRef.get();
            if (addon != null) {
                return addon;
            }
            Class<?> clazz = object.getClass();
            while (!mTypesMap.containsKey(clazz)) {
                if (clazz == Object.class) {
                    // Nothing was found
                    return null;
                }
                clazz = clazz.getSuperclass();
            }
            addon = ((Class<V>) mTypesMap.get(clazz)).newInstance();
            addon.attach(object);
            mStatesMap.put(object, new WeakReference<V>(addon));
            return addon;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
```

**Comment by Prototik** [comment]
Fixed in #436 (3f5a8cf)
